sdk/storage/azure-storage-blob-cryptography/src/main/java/com/azure/storage/blob/cryptography/BlobDecryptionPolicy.java

```diff
@@ -0,0 +1,489 @@
+                boolean padding = encryptedRange.toBlobRange().getOffset()
+                    + encryptedRange.toBlobRange().getCount() > (blobSize(responseHeaders) - ENCRYPTION_BLOCK_SIZE);
+                Metadata metadata = extractMetadataFromResponse(responseHeaders);
+                Flux<ByteBuffer> plainTextData = this.decryptBlob(metadata,
+                    httpResponse.getBody(), encryptedRange, padding);
+
+                return Mono.just(new BlobDecryptionPolicy.DecryptedResponse(httpResponse, plainTextData));
+            } else {
+                return Mono.just(httpResponse);
+            }
+        });
+    }
+
+    /**
+     * Decrypted all or part of an encrypted Block-, Page- or AppendBlob.
+     *
+     * @param metadata The Blob's {@link com.azure.storage.blob.models.Metadata}
+     * @param encryptedFlux The encrypted Flux of ByteBuffer to decrypt
+     * @param encryptedBlobRange A {@link EncryptedBlobRange} indicating the range to decrypt
+     * @param padding Boolean indicating if the padding mode should be set or not.
+     *
+     * @return A Flux ByteBuffer that has been decrypted
+     */
+    Flux<ByteBuffer> decryptBlob(Map<String, String> metadata, Flux<ByteBuffer> encryptedFlux,
+        EncryptedBlobRange encryptedBlobRange, boolean padding) {
+        EncryptionData encryptionData = getAndValidateEncryptionData(metadata);
+
+        // The number of bytes we have put into the Cipher so far.
+        AtomicLong totalInputBytes = new AtomicLong(0);
+        // The number of bytes that have been sent to the downstream so far.
+        AtomicLong totalOutputBytes = new AtomicLong(0);
+
+        return getKeyEncryptionKey(encryptionData)
+            .flatMapMany(contentEncryptionKey -> {
+
+                /*
+                Calculate the IV.
+
+                If we are starting at the beginning, we can grab the IV from the encryptionData. Otherwise,
+                Rx makes it difficult to grab the first 16 bytes of data to pass as an IV to the cipher.
+                As a work around, we initialize the cipher with a garbage IV (empty byte array) and attempt to
+                decrypt the first 16 bytes (the actual IV for the relevant data). We throw away this "decrypted"
+                data. Now, though, because each block of 16 is used as the IV for the next, the original 16 bytes
+                of downloaded data are in position to be used as the IV for the data actually requested and we are
+                in the desired state.
+                 */
+                byte[] iv = new byte[ENCRYPTION_BLOCK_SIZE];
+                /*
+                Adjusting the range by <= 16 means we only adjusted to align on an encryption block boundary
+                (padding will add 1-16 bytes as it will prefer to pad 16 bytes instead of 0) and therefore the key
+                is in the metadata.
+                 */
+                if (encryptedBlobRange.offsetAdjustment() <= ENCRYPTION_BLOCK_SIZE) {
```

Comment:
This could be made into an if/else where the else creates the empty byte array?

Reply:
It's true. Up to you which you prefer. 

---
In reply to: [328298432](https://github.com/Azure/azure-sdk-for-java/pull/5528#discussion_r328298432) [](ancestors = 328298432)